File: courses/codes/HW6Q1.java
import java.util.*;

public class HW6Q1 {
    final static Set<Character> VOWELS = Set.of('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U');

    public static String lastSyllable(String word) {
        for (int i = word.length() - 1; i >= 0; --i) {
            if (VOWELS.contains(word.charAt(i))) {
                return word.substring(i).toLowerCase();
            }
        }
        return word.toLowerCase();
    }

    public static void main(String args[]) {
        Scanner scan = new Scanner(System.in);
        int n = Integer.parseInt(scan.nextLine());
        for (int i = 0; i < n; ++i) {
            String[] poem = new String[4];
            for (int j = 0; j < 4; ++j) {
            	String[] line = scan.nextLine().split(" ");
                poem[j] = lastSyllable(line[line.length-1]);
            }
            boolean s0s1 = poem[0].equals(poem[1]);
            boolean s0s2 = poem[0].equals(poem[2]);
            boolean s0s3 = poem[0].equals(poem[3]);
            boolean s1s2 = poem[1].equals(poem[2]);
            boolean s1s3 = poem[1].equals(poem[3]);
            boolean s2s3 = poem[2].equals(poem[3]);

            if (s0s1 && s0s2 && s0s3) {
                System.out.println("perfect");
            } else if (s0s1 && s2s3) {
                System.out.println("even");
            } else if (s0s2 && s1s3) {
                System.out.println("cross");
            } else if (s0s3 && s1s2) {
                System.out.println("shell");
            } else {
                System.out.println("free");
            }
        }
        scan.close();
    }
}

File: courses/codes/python-class/hw6_q1.py
vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}


def last_syllable(word):
    for i in range(len(word)-1, -1, -1):
        if word[i] in vowels:
            return word[i:len(word)].lower();
    return word.lower()


n = int(input())
for i in range(n):
    poem = [last_syllable(input().split()[-1]) for j in range(4)]
    if poem[0] == poem[1] == poem[2] == poem[3]:
        print("perfect")
    elif poem[0] == poem[1] and poem[2] == poem[3]:
        print("even")
    elif poem[0] == poem[2] and poem[1] == poem[3]:
        print("cross")
    elif poem[0] == poem[3] and poem[1] == poem[2]:
        print("shell")
    else:
        print("free")

